Wrappers/Python/ccpi/optimisation/algorithms/GradientDescent.py

```diff
@@ -84,10 +96,36 @@ def update(self):
         '''Single iteration'''
         if self.memopt:
             self.objective_function.gradient(self.x, out=self.x_update)
+            
+            if self.update_rate:
+                self.rate = self.armijo_rule() * 2.
+
             self.x_update *= -self.rate
             self.x += self.x_update
         else:
-            self.x += -self.rate * self.objective_function.gradient(self.x)
+            if self.update_rate:
+                self.rate = self.armijo_rule() * 2.
+            
+            self.x_update = self.objective_function.gradient(self.x)
+            self.x += -self.rate * self.x_update
 
     def update_objective(self):
         self.loss.append(self.objective_function(self.x))
+
+    def armijo_rule(self):
+        '''Applies the Armijo rule to calculate the step size (rate)
+
+        https://projecteuclid.org/download/pdf_1/euclid.pjm/1102995080
+        '''
+        f_x = self.objective_function(self.x)
+        if not hasattr(self, 'x_update'):
+            self.x_update = self.objective_function.gradient(self.x)
+        while True:
+            f_x_a = self.objective_function(self.x - self.alpha * self.x_update)
+            sqnorm = self.x_update.squared_norm()
+            if f_x_a - f_x <= - ( self.alpha/2. ) * sqnorm:
+                break
+            else:
+                self.k += 1.
+                self.alpha = self.alpha / 2.**(self.k)
```

Comment:
I think this is reducing alpha too quickly. It should only divide by 2 each time.